# TASK 4.5.
# Table of Contents
## 1. Import libraries and data sets
## 2. Exercise consistency checks
### 2.1. Practise checking mixed data types
### 2.2. Check missing values
### 2.3. Check duplicates
### 2.4. Export dataframe
## 3. Task consistency checks
### 3.1 Check mixed data types
### 3.2. Check missing values
### 3.3. Check duplicates
### 3.4. Export clean dataframes

---

## 1. Import libraries and data sets

In [2]:
# Import libraries
import pandas as pd
import numpy as np
import os

In [4]:
path = r'/Users/mariliamachado/Documents/CareerFoundry/11-2024 Instacart Basket Analysis'

In [6]:
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col = False)

In [8]:
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_wrangled.csv'), index_col = False)

## 2. Exercise consistency checks

### 2.1 Practise checking mixed data types

In [12]:
# Create a data frame
df_test = pd.DataFrame()

In [14]:
# Create a mixed type column
df_test['mix'] = ['a', 'b', 1, True]

In [18]:
df_test.head()


mix
0     a
1     b
2     1
3  True

In [27]:
# Check for mixed types
for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


/var/folders/1l/dnjb8fld07b4wz73yz4dyxqw0000gn/T/ipykernel_76867/772068622.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)


### 2.2. Check missing values

In [81]:
# Find missing values
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [143]:
# Subset with the missing values
df_nan = df_prods[df_prods['product_name'].isnull() == True]


In [145]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [47]:
df_prods.shape

(49693, 5)

In [147]:
# Create new clean dataframe
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [149]:
# Check number of rows/columns
df_prods_clean.shape 

(49677, 5)

### 2.3. Check duplicates

In [151]:
# Check duplicates
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [55]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [153]:
# Check number of rows/columns
df_prods_clean.shape

(49677, 5)

In [155]:
# Remove duplicates
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()


In [69]:
# Check number of rows/columns
df_prods_clean_no_dups.shape

(49672, 5)

### 2.4. Export dataframe

In [157]:
# Export clean dataframe
df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data','Prepared Data', 'products_checked.csv'))

## 3. Task consistency checks

2. Run the df.describe() function on your df_ords dataframe. Using your new knowledge about how to interpret the output of this function, share in a markdown cell whether anything about the data looks off or should be investigated further.
Tip: Keep an eye on min and max values!

In [95]:
df_ords.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       orders_day_of_week  order_time_hour  days_since_prior_order  
count        3.421083e+06     3.421083e+06            3.214874e+06  
mean         2.776219e+00     1.345202e+01            1.111484e+01  
std          2.046829e+00     4.226088e+00            9.206737e+00  
min          0.000000e+00     0.000000e+00            0.000000e+00  
25%          1.000000e+00     1.000000e+01            4.000000e+00  
50%          3.000000e+00     1.300000e+01            7.000000e+00  
75%          5.000000e+00     1.600000e+01            1.500000e+01  
max          6.000000e+00     2.300000e+01            3.000000e+01

There is a column that is unnamed, which has the same exact values as order_id, which could mean is a repeated column. Days_since_prior_order column must have missing values, since the count is lower than all the other columns. The min value of orders_day_of_week is 0, which means that they are considering the first day of the week as 0 and not 1. The max value of the same column is 6 instead of 7, since we the values start from 0, even though culturally, we expect it to be 7 days of the week. As for order_time_hour, the min value is 0 (midnight) and the max value is 23h (11 p.m.). Finally, the min values for days_since_prior_order is 0, which makes sense and the max value seems to be 30, which suggests that this value is capped at 1 month.

### 3.1 Check mixed data types

3. Check for mixed-type data in your df_ords dataframe.
4. If you find mixed-type data, fix it. The column in question should contain observations of a single data type.

In [109]:
# Check for mixed types (note that I changed applymap to just map due to an error message I got)
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].map(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)

In [122]:
df_ords.dtypes

Unnamed: 0                  int64
order_id                    int64
user_id                     int64
order_number                int64
orders_day_of_week          int64
order_time_hour             int64
days_since_prior_order    float64
dtype: object

There is no mixed-type data in this dataframe.

5. Run a check for missing values in your df_ords dataframe.
In a markdown cell, report your findings and propose an explanation for any missing values you find.

### 3.2. Check missing values

In [126]:
# Find missing values
df_ords.isnull().sum()

Unnamed: 0                     0
order_id                       0
user_id                        0
order_number                   0
orders_day_of_week             0
order_time_hour                0
days_since_prior_order    206209
dtype: int64

The days_since_prior_order column has 206209 missing values. First, I thought that maybe these were not missing values, but just the number 0 meaning there were 0 days since the last order, but I realized that other columns also have the possibility of having 0 as a value and they were not flagged as missing in the previous command. An explanation is that maybe the clients' first order have a NA value, since that column is not applicable to them.

6. Address the missing values using an appropriate method.
In a markdown cell, explain why you used your method of choice.

In [139]:
# Check the missing values by creating subset
df_ords_nan = df_ords[df_ords['days_since_prior_order'].isnull() == True]


In [175]:
df_ords_nan

Unnamed: 0  order_id  user_id  order_number  orders_day_of_week  \
0                 0   2539329        1             1                   2   
11               11   2168274        2             1                   2   
26               26   1374495        3             1                   1   
39               39   3343014        4             1                   6   
45               45   2717275        5             1                   3   
...             ...       ...      ...           ...                 ...   
3420930     3420930    969311   206205             1                   4   
3420934     3420934   3189322   206206             1                   3   
3421002     3421002   2166133   206207             1                   6   
3421019     3421019   2227043   206208             1                   1   
3421069     3421069   3154581   206209             1                   3   

         order_time_hour  days_since_prior_order  
0                      8                     NaN  
11                    11                     NaN  
26                    14                     NaN  
39                    11                     NaN  
45                    12                     NaN  
...                  ...                     ...  
3420930               12                     NaN  
3420934               18                     NaN  
3421002               19                     NaN  
3421019               15                     NaN  
3421069               11                     NaN  

[206209 rows x 7 columns]

In [183]:
# Check the frequency of order number in our subset
df_ords_nan['order_number'].value_counts()

order_number
1    206209
Name: count, dtype: int64

In [185]:
# Check the frequency of order_number = 1 in df_ords
df_ords['order_number'].value_counts()

order_number
1      206209
2      206209
3      206209
4      206209
5      182223
        ...  
96       1592
97       1525
98       1471
99       1421
100      1374
Name: count, Length: 100, dtype: int64

It seems that my assumption of the missing values referring to the first order is correct.

In [196]:
# I will create a new column that can let us know when it is a first order
df_ords['first_order'] = df_ords['days_since_prior_order'].isna()

In [200]:
# Check if it worked
df_ords

Unnamed: 0  order_id  user_id  order_number  orders_day_of_week  \
0                 0   2539329        1             1                   2   
1                 1   2398795        1             2                   3   
2                 2    473747        1             3                   3   
3                 3   2254736        1             4                   4   
4                 4    431534        1             5                   4   
...             ...       ...      ...           ...                 ...   
3421078     3421078   2266710   206209            10                   5   
3421079     3421079   1854736   206209            11                   4   
3421080     3421080    626363   206209            12                   1   
3421081     3421081   2977660   206209            13                   1   
3421082     3421082    272231   206209            14                   6   

         order_time_hour  days_since_prior_order  first_order  
0                      8                     NaN         True  
1                      7                    15.0        False  
2                     12                    21.0        False  
3                      7                    29.0        False  
4                     15                    28.0        False  
...                  ...                     ...          ...  
3421078               18                    29.0        False  
3421079               10                    30.0        False  
3421080               12                    18.0        False  
3421081               12                     7.0        False  
3421082               14                    30.0        False  

[3421083 rows x 8 columns]

I believe those missing values are important because they are not a mistake, they are correct since if it's a first order, there is no available data about the days since a prior order. Therefore, I created another column letting us know this information.

### 3.3. Check duplicates

7. Run a check for duplicate values in your df_ords data.
In a markdown cell, report your findings and propose an explanation for any duplicate values you find.
8. Address the duplicates using an appropriate method.
In a markdown cell, explain why you used your method of choice.

In [206]:
# Check duplicates
df_ords_dups = df_ords[df_ords.duplicated()]

In [208]:
df_ords_dups

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, order_number, orders_day_of_week, order_time_hour, days_since_prior_order, first_order]
Index: []

It seems there are no duplicates and according to previous steps, my assumption that the Unnamed column was a duplicate of order_id was wrong.

### 3.4. Export clean dataframes

9. Export your final, cleaned df_prods and df_ords data as “.csv” files in your “Prepared Data” folder and give them appropriate, succinct names.

In [219]:
# Export clean df_prods dataframe
df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data','Prepared Data', 'products_checked.csv'))

In [222]:
# Export clean df_ords dataframe
df_ords.to_csv(os.path.join(path, '02 Data','Prepared Data', 'ords_checked.csv'))